### ISSUE TITLE
Support parsing symbols from the linking section for wasm objects

### ISSUE THREAD

**Issue by bjorn3** [gimli-rs/object#471]
In case of wasm object files there is no export section. Instead all exported symbols are defined in the "linking" section. This is necessary for https://github.com/rust-lang/rust/pull/97485. See https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md for the full specification.

**Comment by philipc** [comment]
Is this blocked on wasmparser support? Can you provide a wasm file that I can use for testing this?

**Comment by bjorn3** [comment]
```
cat > foo.rs <<EOF
#[no_mangle]
extern "C" fn foo() -> u8 { bar() }
fn bar() -> u8 { *FOO.lock().unwrap() }
static FOO: std::sync::Mutex<u8> = std::sync::Mutex::new(42);
EOF
rustc foo.rs --target wasm32-unknown-unknown --crate-type lib --emit obj
```

produces [foo.o.gz](https://github.com/gimli-rs/object/files/11123968/foo.o.gz) (gzip compressed without a tar archive as github didn't accept a gzipped tarball despite saying that `.tgz` is accepted as file extension)

**Comment by bjorn3** [comment]
This can be implemented without wasmparser changes.

**Comment by philipc** [comment]
> This can be implemented without wasmparser changes.

I didn't see anything in wasmparser that can parse these sections. Are you recommending we implement our own parsing? If we're going to use wasmparser for parsing wasm then I expect parsing of the linking section to be implemented in wasmparser.

**Comment by bjorn3** [comment]
Wasmparser indeed doesn't have any code to parse this AFAIK. Adding the code to wasmparser would avoid code duplication with anyone else who needs it, but it would be possible to put the parse code for the linking custom section in object too. Wasmparser gives a raw byte slice as content for custom sections.

**Comment by philipc** [comment]
If someone has to write the code, they may as well do it in wasmparser, so let's wait for someone to do that.

**Comment by andylizi** [comment]
I think this is now added in bytecodealliance/wasm-tools#1348

**Comment by tgross35** [comment]
Does this explain why wasm symbols sometimes show up with no name? For a small demo:

```rust
// rustc a.rs --emit=obj --target=wasm32-unknown-unknown --crate-type=lib -Cdebuginfo=0

extern "C" { fn some_undef(); }

#[unsafe(no_mangle)]
extern "C" fn somefunc() {
    unsafe { some_undef() };
}
```

llvm-nm correctly shows the two symbols but using `object`, `somefunc` shows up as an empty name.

```
$ llvm-nm a.o
         U some_undef
00000001 T somefunc
./util a.o
Symbol { name: "env", address: 0, size: 0, kind: File, section: None, scope: Dynamic, weak: false, flags: None }
Symbol { name: "__linear_memory", address: 0, size: 0, kind: Data, section: Undefined, scope: Dynamic, weak: false, flags: None }
Symbol { name: "some_undef", address: 0, size: 0, kind: Text, section: Undefined, scope: Dynamic, weak: false, flags: None }
Symbol { name: "", address: 0, size: 0, kind: File, section: None, scope: Compilation, weak: false, flags: None }
Symbol { name: "", address: 2, size: 9, kind: Text, section: Section(SectionIndex(10)), scope: Compilation, weak: false, flags: None }
```

Util for reference

```rust
fn main() {
    let path = env::args().nth(1).unwrap();
    let data = fs::read(path).expect("reading file failed");
    let obj = object::File::parse(&*data).expect("failed to parse object");
    for sym in obj.symbols() {
        println!("{sym:?}");
    }
}
```

**Comment by philipc** [comment]
Yes, that's a Wasm object file so this issue needs to be fixed to properly support it.

**Comment by philipc** [comment]
I've started implementing this.

**Comment by philipc** [comment]
Can someone who wants this please review #766.

**Comment by philipc** [comment]
Symbol names should be correct now. There is further work to do:
- determine the section and address for non-code symbols
- handle other linking subsections, in particular relocations, but also segments and comdat sections might be useful
- not directly related to the linking section, but we should stop returning synthetic symbols for imports and exports. We should only return symbols that actually exist.
